### ISSUE TITLE
Bessel functions in JAX

### ISSUE THREAD

**Issue by shashankdholakia** [JAXtronomy/spexial#15]
Hi everyone,

This looks like a great and much needed package handling special functions in JAX! We've been in an [ongoing discussion](https://github.com/jax-ml/jax/pull/17038) with the JAX team about Bessel functions in Jax as the jax.scipy.special bessel functions suffer from numerical instabilities and diverge from the syntax of the equivalent scipy functions. Likewise, having numerically stable spherical bessel functions are important for a wide range of astronomical applications. Would you consider hosting our implementations of [bessel](https://github.com/jax-ml/jax/pull/17038) and [spherical bessel](https://github.com/shashankdholakia/harmonix/blob/main/src/harmonix/utils.py) functions?

**Comment by adrn** [comment]
In principle, yes! But reading through the PR discussion it does seem reasonable to raise a warning in these functions if x64 is not enabled.

**Comment by nstarman** [comment]
Status update on where `spexial` stands, since the offer here is still open and it's worth being precise about what would and wouldn't overlap.

## What we ship today

Only one Bessel family: the **modified Bessel function of the second kind at integer orders 0, 1, 2**, in both plain and exponentially scaled form.

| `spexial` | equivalent | `jax.scipy.special` | derivative |
|---|---|---|---|
| `k0` | `scipy.special.k0` | — | `-k1(z)` |
| `k1` | `scipy.special.k1` | — | `-k0(z) - k1(z)/z` |
| `k2` | `scipy.special.kn(2, z)` | — | `-k1(z) - (2/z) k2(z)` |
| `k0e` | `scipy.special.k0e` | — | `k0e(z) - k1e(z)` |
| `k1e` | `scipy.special.k1e` | — | `k1e(z) - k0e(z) - k1e(z)/z` |
| `k2e` | `scipy.special.kve(2, z)` | — | `k2e(z) - k1e(z) - (2/z) k2e(z)` |

All six carry a hand-written `custom_jvp` rather than differentiating the series, and all `jit`/`vmap`/`grad` cleanly. The uppercase `K0`…`K2e` spellings still work but are deprecated.

Ascending series below `z = 9`, asymptotic expansion above. Measured against SciPy on 20k points in float64:

```
k0  worst 1.9e-07  at z = 8.993   (just below the cross-over)
k1  worst 1.7e-07  at z = 8.993
k2  worst 1.2e-07  at z = 8.993

k0 by region:  [0.001, 1] 8.6e-16   [1, 8.9] 1.3e-07   [9, 15] 8.3e-09
               [15, 30] 2.4e-13     [30, 700] 8.0e-16
```

So the error is concentrated in a narrow band under the cross-over and is near machine precision everywhere else. `k0` underflows to zero above `z = 705.3` because the true value is subnormal there — `k0e` is exact past it (`k0e(1e5) = 0.003963`, matching SciPy).

## What we do not ship

**Nothing from this offer is covered.** Specifically missing:

- `jv`, `yv`, `iv`, `kv` at general (real, non-integer) order — the whole of [jax-ml/jax#17038](https://github.com/jax-ml/jax/pull/17038)
- `yn`, and `kn` at orders other than 0, 1, 2
- every spherical Bessel function — `spherical_jn`, `spherical_yn` and friends

For reference, `jax.scipy.special` today has only `bessel_jn` (integer order, via `v=` keyword and a fixed `n_iter=50`), plus `i0`, `i0e`, `i1`, `i1e`. No `K` at any order, no general order, nothing spherical. SciPy's own `k0` works eagerly on a JAX array by converting to NumPy, but raises `TracerArrayConversionError` under both `jit` and `grad`, so it is not a fallback inside a traced function.

So both halves of the original offer — the general-order set and the spherical set — remain wanted and wouldn't collide with anything we have.

## On @adrn's x64 point

Worth reporting what happened when we hit exactly this in `k0`. Rather than warn, we made the branch cross-over depend on the dtype's `eps`:

The ascending series evaluates `-(log(z/2) + γ) I0(z) + Σ(...)`, whose two terms are each ~`e^z/√z` and cancel down to a result of ~`e^-z` — a loss of about `2z/ln(10)` decimal digits. float64 has 16 to spend, so at `z = 9` it still has 8 left. float32 has 7, and at `z = 9` it has **none**: the result came out *negative*. Moving the float32 cross-over to 4.65 caps the worst float32 error at 7e-3 (measured: 6.6e-3 at `z = 4.63`, median 9.6e-8), against 1.9e-7 in float64.

Two things that generalise to whatever lands here:

1. The right cross-over is not the float64 one rescaled — it has to be measured per precision, and keyed on `jnp.finfo(dtype).eps` rather than on a dtype name so that float16/bfloat16 get the conservative branch instead of inheriting a float64 choice.
2. A warning would have been the wrong instrument for this particular failure. It was silently returning a *negative* `K0`, which no amount of warning-reading prevents; the fix had to be in the branch point. Warnings are still reasonable where a function is genuinely unusable in float32, but it's worth checking first whether the cross-over can simply be moved.

Happy to take a PR against any of the missing families — the `k*` module is a reasonable template for the accuracy documentation and `custom_jvp` conventions we've settled on.

<sub>Measured on jax 0.10.2, CPU, against SciPy; `spexial` at 9e38f92.</sub>

**Comment by NeilGirdhar** [comment]
Cool!  Here's a start for a couple in case anyone wants to work on it:
```py
def bessel_iv_ratio(v: jax.array, x: jax.array, /, *, iterations: int = 200) -> jax.array:
    """return ``i_v(x) / i_(v - 1)(x)`` for nonnegative real ``x``.

    the implementation uses a backward recurrence, which is stable for the ratio and remains fully
    differentiable with respect to both ``v`` and ``x``.
    """
    v, x = jnp.broadcast_arrays(jnp.asarray(v), jnp.asarray(x))
    q = jnp.zeros_like(v + x)

    def body(i: int, q: jax.array) -> jax.array:
        order = v + (iterations - i) - 1.0
        return x / (2.0 * order + x * q)

    return jax.lax.fori_loop(0, iterations, body, q)


def log_bessel_ive(v: jax.array, x: jax.array, /) -> jax.array:
    """return ``log(ive(v, x))`` for nonnegative real ``x``.

    this uses the power series for ``i_v(x)`` in log-space and then subtracts ``x`` to obtain the
    exponentially scaled quantity. the implementation is pure jax and remains differentiable with
    respect to both ``v`` and ``x``.
    """
    v, x = jnp.broadcast_arrays(jnp.asarray(v), jnp.asarray(x))
    dtype = jnp.result_type(v, x, 1.0)
    v = v.astype(dtype)
    x = x.astype(dtype)
    safe_x = jnp.maximum(x, jnp.finfo(dtype).tiny)
    terms = jnp.arange(200, dtype=dtype).reshape((1,) * x.ndim + (200,))
    v_expanded = v[..., none]
    log_terms = (
        (2.0 * terms + v_expanded) * jnp.log(safe_x[..., none] / 2.0)
        - gammaln(terms + 1.0)
        - gammaln(terms + v_expanded + 1.0)
    )
    retval = logsumexp(log_terms, axis=-1) - x
    return jnp.where(x == 0, jnp.where(v == 0, 0.0, -jnp.inf), retval)
```

No hand-written pullback function though.
